[net-diag] define child and Child IPv6 Address List TLV

abtink opened the pull request:
This commit contains multiple changes to Network Diagnostics modules.

It adds three new TLVs: Child TLV, Child IPv6 Address List TLV, and Neighbor TLV. Child TLV contains information about a child entry including RLOC16, extended MAC address, mode (rx-on-idle, FTD/MTD, full netdata, CSL-sync), timeout, age, supervision interval, CSL period and timeout, and link quality info such as RSS (last and average) and frame/message error rates. Neighbor TLV provides similar info for a router neighbor. The Child IPv6 Address List TLV allows to query a parent to get the list of IPv6 addresses registered by its MTD children.

The new TLVs can be requested in Network Diagnostic Get Query "d/dq" message to a router (unicast) or multiple routers (multicast). The router responds with one or more Network Diagnostic Get Answer "d/da" messages.

We allow multiple Answer messages to be sent in response to a Query. This is to support the case where a router has many children, and we want to avoid sending a large message that contains all of the TLVs for all of the children. Instead, we can send a sequence of Answer messages, each of which contains a subset of the TLVs.

Each Answer message includes an Answer TLV that indicates the index of the message in the sequence. The first Answer message has an index of 0, the second has an index of 1, and so on. The Answer TLV also indicates whether the message is the last one in the sequence, or if there are more Answer messages to come. Answer messages are sent in sequence, and the next one is not sent until we receive a successful CoAP ACK/response for the previous one.

This commit adds a Query ID TLV (with a 16-bit identifier as it value) that can be optionally included in a Query message. The Query ID is echoed back in all Answer messages associated with the same Query message. This allows a node that sends multiple Queries to distinguish between the received Answers.

This commit updates `MeshDiag` to use the newly added TLVs to query the child table or children IPv6 addresses of a router. New public OpenThread APIs under `otMeshDiag` are added for this, along with CLI commands and their documentation. A test-case is added to validate the newly added mechanism.

----

Related to [SPEC-1151](https://threadgroup.atlassian.net/browse/SPEC-1151) and [SPEC-1150](https://threadgroup.atlassian.net/browse/SPEC-1150).

EskoDijk reviewed (changes requested):
Provided some comments inline.
In general, would it be desired to keep this functionality out of a Thread 1.1 build and a 1.2 build? Since it's 1.3 functionality. If we include it in 1.1 builds then I wonder what the exact meaning of the "1.1" build flag has become.

abtink commented on include/openthread/mesh_diag.h:
```diff
@@ -221,6 +227,70 @@ otError otMeshDiagGetNextIp6Address(otMeshDiagIp6AddrIterator *aIterator, otIp6A
+    uint16_t     mSupervisionInterval; ///< Supervision interval in seconds. Zero to indicate not used.
+    int8_t       mAverageRssi;         ///< Average RSSI.
+    int8_t       mLastRssi;            ///< RSSI of last received frame.
+    uint16_t     mFrameErrorRate;      ///< Frame error rate (0xffff->100%). Zero if not known or supported.
```
Yes. Thanks. I agree `Zero if not known` can be confusing. I intended it as if you don't support error rate track, set then to zero, but we can use `mSupportsErrRate` flag to tell if the error rate values are valid, otherwise value should not matter and we should ignore them.

In the next push, I remove this and make it simpler. Also added a comment at the top to explain the error rate a bit more.

abtink commented on include/openthread/mesh_diag.h:
```diff
@@ -221,6 +227,70 @@ otError otMeshDiagGetNextIp6Address(otMeshDiagIp6AddrIterator *aIterator, otIp6A
+    uint16_t     mMessageErrorRate;    ///< (IPv6) msg error rate (0xffff->100%). Zero if not known or supported.
+    uint16_t     mQueuedMessageCount;  ///< Number of queued message for indirect tx to child.
+    uint16_t     mCslPeriod;           ///< CSL Period in unit of 10 symbols. Zero indicated disabled.
+    uint32_t     mCslTimeout;          ///< CSL Timeout in seconds. Zero if not suppurated.
```
Yes. I am going to remove this.

abtink commented on src/core/thread/network_diagnostic_tlvs.hpp:
```diff
@@ -615,6 +617,177 @@ class TypeListTlv : public Tlv, public TlvInfo<Tlv::kTypeList>
+     * `kFlagsTrackErrRate` from `GetFlags()` indicates whether or not the implementation supports tracking of error
+     * rates and whether or not the value in this field is valid.
+     *
+     * @returns The Message Error Rate (0xffff->100%). Zero if not known or supported.
```
Good idea, will add it.

EskoDijk commented on src/cli/cli.cpp:
```diff
@@ -5807,6 +5807,85 @@ template <> otError Interpreter::Process<Cmd("meshdiag")>(Arg aArgs[])
+        error = OT_ERROR_PENDING;
+    }
+    /**
+     * @cli meshdiag childrenip6addrs
```
Possible easier and shorter command name is `childip6` ?

abtink replied:
Good idea. Will rename it in next push. I will add more to this PR and will add "Answer TLV" for tracking multiple answers for the same query and will add a "Query ID TLV" so we can associate a query with its answer(s)`. It makes this PR bigger but I feel make sense to add and test all these related mechanisms together.

----

@EskoDijk want to also thank you for reviewing the PRs and specially paying attention to details and giving good suggestions. I do appreciate it. Thank you 😃 🙏

EskoDijk replied:
@abtink Thanks! One thing still to consider is how would a Router with many children handle the multiple Answer messages? Frame buffer / queue space is limited and I think the Thread spec text prescribes the device to apply "backpressure" i.e. stop sending further Answers if the frame queue reaches a certain size or frame age (? not sure).

Ideally it would be best if the client would be in control using a request/response type protocol, as e.g. CoAP blockwise is also using - each next block is explicitly fetched by the client. That also reduces DoS opportunities e.g. sending a small request frame could cause an avalanche of data on the mesh.

Further thoughts: Also in CoAP blockwise, instead of a "Query ID TLV" the Token value is used as the identifier of the transaction. Not sure if we really need to reinvent all such mechanisms here?   The main problem of "Blockwise" is that it assumes one data blob is fetched and that data can be arbitrarily divided into blocks of certain sizes (32/64/128 etc).  That's not compatible with our model here.

abtink replied:
> One thing still to consider is how would a Router with many children handle the multiple Answer messages? Frame buffer / queue space is limited and I think the Thread spec text prescribes the device to apply "backpressure" i.e. stop sending further Answers if the frame queue reaches a certain size or frame age (? not sure).

Two ideas on this:
- First, we can rely on `Coap` module itself since it keeps the messages to retry if they fail, so if msg is dropped due to back-pressure logic, it will be retried by `Coap` module.
- Second one which I want to try out is to handle this in `NetworkDiag::Server` module itself: It sends the first answer and then waits for CoAP ack for it before sending the next answer (stop sending answer if no ack for a previous answer). I like this better for sure (I think it requires some changes in `Coap` module to add a callback for "AckRx" to use with confirmable CoAP msgs).

abtink replied:
> Also in CoAP blockwise, instead of a "Query ID TLV" the Token value is used as the identifier of the transaction. Not sure if we really need to reinvent all such mechanisms here? The main problem of "Blockwise" is that it assumes one data blob is fetched and that data can be arbitrarily divided into blocks of certain sizes (32/64/128 etc). That's not compatible with our model here.

I think adding "Query ID TLV" can help with multicast queries as well (same node sending multiple queries). 
We cannot use CoAP block-wise for query/answer model. We have an implementation of block-wise transfer in Coap, but it is not really enabled on devices and/or tested much (it is intended for experimental use by application developers). It adds some good chunk of code overhead (~5-6 KB of code when enabled)

EskoDijk replied:
Agree, something specific in this case would be better than forcing CoAP Blockwise (which doesn't fit our use case very well).

On the 2nd idea of waiting for the CoAP ACK: this Ack, in response to a TMF Answer, is in fact (or should  be) a "piggybacked response" which contains the Ack with the 2.04 response. The latter (response code) is for sure something that the CoAP module ought to expose to the next higher layer. Because it is really an application-level concept.  So at CoAP module level it would make perfect sense to expose this response code and use a callback or so to get its value back to the layer using CoAP.  

Exposing the ACK itself in this way would make less sense, I think, this is really a CoAP-internal layer used for the reliability.  At least normally the application code doesn't need to be informed of all the details of retransmissions or notifying the application that the request was received - which the Ack is about.  But because of the piggybacked communication pattern the Ack and the REsponse tend to get named interchangably, but in principle these are different concepts.

One added benefit of the diagnostic layer getting back the CoAP response code from the entity requesting the data is that it can be used for control of the remainder of Answers:  for example if the requesting entity uses 2.xx success response then the Router can continue sending more Answers. If the entity uses an error code (4.xx / 5.xx) then this may signal the Router to stop sending further Answers.

EskoDijk replied:
Section 10.4.4 of the spec btw lists the current CoAP response codes usable for TMF.  In principle these can equally be used for particular error cases instead of the "dummy 2.04" response we currently define for the TMF Answer.

EskoDijk replied:
@abtink In CoAP, you would always know which CoAP request a CoAP response belongs to: the Token value. That means if a node is sending multiple TMF Answer messages (=CoAP requests) at the same time and to the same destination, it can still distinguish the different CoAP responses (e.g. all 2.04) coming in by their Token value.

But maybe the problem you indicate is one layer above CoAP? E.g. if the application layer doesn't see the CoAP Token because the CoAP layer adds it, then the application cannot match requests and responses using the Token value. Then the API for the "send CoAP request" call could immediately return a handle (ID, or pointer) to identify this request and the same handle is used for responses coming in.  Or the other solution of a context which achieves the same I think.

abtink replied:
> But maybe the problem you indicate is one layer above CoAP?
>  Or the other solution of a context which achieves the same I think.

Yes Exactly. :). I was searching for a way to realize this without needing to track anything in `NetDiag` module and/or need to change the existing CoAP methods and callbacks.

abtink commented:
Pushed a new version of this PR. New changes are:
- Adding "Answer TLV" to track index of Answer
- Adding "Query ID TLV"
- Mechanism to send Answer messages one by one (wait for the CoAP response/ack for previous one before sending the next one)
- Relax/Fix some of the test script (some where strict in the sense that expected to see exact set of TLVs).

The [commit log and PR description](https://github.com/openthread/openthread/pull/8866#issue-1624149209) is also updated with more details.

abtink commented:
Changes in new push:
- Added support for `mConnectionTime` (seconds since link establishment with child/neighbors) in Child and Neighbor TLVs (support for this was added in https://github.com/openthread/openthread/pull/9007).
- Added support for connection time in mesh-diag and the related CLI commands.
- Addressed suggestion from @Vyrastas posted on https://github.com/openthread/openthread/pull/9024.

abtink commented:
New push to this PR:
- It now include the MLE counters TLV from https://github.com/openthread/openthread/pull/9024. 
- Renamed to "Router Neighbor TLV" from "Neighbor TLV" (make it clear it is the routers)
- Removed the device mode flags from "Router Neighbor TLV" (and related `MeshDiag` type).

abtink commented:
In new push, rebased on `main`, addressed conflicts. 
- Used new `Tlv` helper method to simplify processing of TLVs in the received message.
- Updated the function/method APIs to remove "This function/method/type ..".

jwhui commented on src/core/utils/mesh_diag.hpp:
```diff
@@ -164,17 +174,66 @@ class MeshDiag : public InstanceLocator
-     * Cancels an ongoing topology discovery if there one, otherwise no action.
+     * Starts query for child table for a given router.
+     *
+     * @param[in] aRouter16        The RLOC16 of router to query.
```
```suggestion
     * @param[in] aRloc16          The RLOC16 of router to query.
```

jwhui commented on src/core/utils/mesh_diag.hpp:
```diff
@@ -164,17 +174,66 @@ class MeshDiag : public InstanceLocator
+    /**
+     * Sends a query to a parent to retrieve the IPv6 addresses of all its MTD children.
+     *
+     * @param[in] aRouter16        The RLOC16 of parent to query.
```
```suggestion
     * @param[in] aRloc16          The RLOC16 of parent to query.
```

jwhui commented on src/core/utils/mesh_diag.hpp:
```diff
@@ -164,17 +174,66 @@ class MeshDiag : public InstanceLocator
+    /**
+     * Starts query for router neighbor table for a given router.
+     *
+     * @param[in] aRouter16        The RLOC16 of router to query.
```
```suggestion
     * @param[in] aRloc16          The RLOC16 of router to query.
```

jwhui commented on include/openthread/mesh_diag.h:
```diff
@@ -82,7 +82,7 @@ typedef struct otMeshDiagIp6AddrIterator otMeshDiagIp6AddrIterator;
 
 /**
- * This constant indicates that Thread Version is unknown.
+ * Specified that Thread Version is unknown.
```
```suggestion
 * Specifies that Thread Version is unknown.
```

jwhui commented on include/openthread/mesh_diag.h:
```diff
@@ -221,6 +229,181 @@ otError otMeshDiagGetNextIp6Address(otMeshDiagIp6AddrIterator *aIterator, otIp6A
+ * Starts query for child table for a given router.
+ *
+ * @param[in] aInstance        The OpenThread instance.
+ * @param[in] aRouter16        The RLOC16 of router to query.
```
```suggestion
 * @param[in] aRloc16.         The RLOC16 of router to query.
```

jwhui commented on include/openthread/mesh_diag.h:
```diff
@@ -221,6 +229,181 @@ otError otMeshDiagGetNextIp6Address(otMeshDiagIp6AddrIterator *aIterator, otIp6A
+ * Sends a query to a parent to retrieve the IPv6 addresses of all its MTD children.
+ *
+ * @param[in] aInstance        The OpenThread instance.
+ * @param[in] aRouter16        The RLOC16 of parent to query.
```
```suggestion
 * @param[in] aRloc16          The RLOC16 of parent to query.
```

jwhui commented on include/openthread/mesh_diag.h:
```diff
@@ -221,6 +229,181 @@ otError otMeshDiagGetNextIp6Address(otMeshDiagIp6AddrIterator *aIterator, otIp6A
+ * @param[in] aError       OT_ERROR_PENDING            Indicates there are more entries in the table.
+ *                         OT_ERROR_NONE               Indicates the table is finished.
+ *                         OT_ERROR_RESPONSE_TIMEOUT   Timed out waiting for response.
+ * @param[in] aChildEntry  The neighbor entry (can be null if `aError` is OT_ERROR_RESPONSE_TIMEOUT or OT_ERROR_NONE).
```
```suggestion
 * @param[in] aNeighborEntry  The neighbor entry (can be null if `aError` is OT_ERROR_RESPONSE_TIMEOUT or OT_ERROR_NONE).
```

jwhui commented on include/openthread/mesh_diag.h:
```diff
@@ -221,6 +229,181 @@ otError otMeshDiagGetNextIp6Address(otMeshDiagIp6AddrIterator *aIterator, otIp6A
+ * Starts query for router neighbor table for a given router.
+ *
+ * @param[in] aInstance        The OpenThread instance.
+ * @param[in] aRouter16        The RLOC16 of router to query.
```
```suggestion
 * @param[in] aRloc16          The RLOC16 of router to query.
```

jwhui commented on include/openthread/netdiag.h:
```diff
@@ -201,6 +208,29 @@ typedef struct otNetworkDiagMacCounters
+    uint16_t mPartitionIdChanges;            ///< Number of changes to partition ID.
+    uint16_t mBetterPartitionAttachAttempts; ///< Number of attempts to attach to a better partition.
+    uint16_t mParentChanges;                 ///< Number of time device changed its parent.
+    uint64_t mTrackedTime;                   ///< Milliseconds tracked by next counters (zero if not supported)
```
```suggestion
    uint64_t mTrackedTime;                   ///< Milliseconds tracked by next counters (zero if not supported).
```

jwhui replied:
Is this expected to equal the sum of the following counters?

abtink replied:
Yes. It is sort of redundant but mirrors the definition of `otMleCounters`.

jwhui commented on src/core/thread/network_diagnostic.cpp:
```diff
@@ -416,18 +413,378 @@ void Server::HandleTmf<kUriDiagnosticGetQuery>(Coap::Message &aMessage, const Ip
+
+        SuccessOrExit(error = aRequest.Read(offset, tlvType));
+
+        if (tlvType == ChildTlv::kType)
```
Use switch statement?

abtink replied:
Good suggestion. Thanks.

jwhui commented on src/core/thread/network_diagnostic.hpp:
```diff
@@ -153,10 +177,35 @@ class Server : public InstanceLocator, private NonCopyable
+    void        HandleAnswerResponse(Coap::Message          &aNextAnswer,
+                                     Coap::Message          *aResponse,
+                                     const Ip6::MessageInfo *aMessageInfo,
+                                     Error                   aResult);
+
+#endif
```
```suggestion
                                     Error                   aResult);
#endif
```

jwhui commented on src/core/thread/network_diagnostic_tlvs.hpp:
```diff
@@ -224,6 +231,18 @@ typedef StringTlvInfo<Tlv::kVendorSwVersion, Tlv::kMaxVendorSwVersionLength> Ven
 typedef StringTlvInfo<Tlv::kThreadStackVersion, Tlv::kMaxThreadStackVersionLength> ThreadStackVersionTlv;
 
+/**
+ * This class defines Child IPv6 Address List TLV constants and types.
```
```suggestion
 * Child IPv6 Address List TLV constants and types.
```

jwhui commented on src/core/thread/network_diagnostic_tlvs.hpp:
```diff
@@ -224,6 +231,18 @@ typedef StringTlvInfo<Tlv::kVendorSwVersion, Tlv::kMaxVendorSwVersionLength> Ven
+typedef TlvInfo<Tlv::kChildIp6AddressList> ChildIp6AddressListTlv;
+
+/**
+ * This class defines Query ID TLV constants and types.
```
```suggestion
 * Query ID TLV constants and types.
```

jwhui commented on src/core/thread/network_diagnostic_tlvs.hpp:
```diff
@@ -667,6 +686,465 @@ class TypeListTlv : public Tlv, public TlvInfo<Tlv::kTypeList>
+#if OPENTHREAD_FTD
+
+/**
+ * This class implements Child TLV generation and parsing.
```
```suggestion
 * Child TLV generation and parsing.
```

jwhui commented on src/core/thread/network_diagnostic_tlvs.hpp:
```diff
@@ -667,6 +686,465 @@ class TypeListTlv : public Tlv, public TlvInfo<Tlv::kTypeList>
+    static constexpr uint8_t kFlagsTrackErrRate = 1 << 3; ///< Supports tracking error rates.
+
+    /**
+     * This method initializes the TLV using information from a given `Child`.
```
Remove "This method " here and below.

Vyrastas commented on src/cli/cli.cpp:
```diff
@@ -5018,6 +5018,129 @@ template <> otError Interpreter::Process<Cmd("meshdiag")>(Arg aArgs[])
+     * - Device Mode
+     * - RSS (average and last)
+     * - Error rates: frame tx (at MAC layer), IPv6 message tx (above MAC)
+     * - Connection time (seconds since link establishment {dd}d.{hh}:{mm}:{ss} format)
```
Surround the example format in ` so it's code font on the website

`{dd}d.{hh}:{mm}:{ss}`

Vyrastas commented on src/cli/cli.cpp:
```diff
@@ -5018,6 +5018,129 @@ template <> otError Interpreter::Process<Cmd("meshdiag")>(Arg aArgs[])
+     *  - Thread Version
+     *  - RSS (average and last) and link margin
+     *  - Error rates, frame tx (at MAC layer), IPv6 message tx (above MAC)
+     *  - Connection time (seconds since link establishment {dd}d.{hh}:{mm}:{ss} format)
```
Surround the example format in ` so it's code font on the website

`{dd}d.{hh}:{mm}:{ss}`

Vyrastas commented on src/cli/cli.cpp:
```diff
@@ -5018,6 +5018,129 @@ template <> otError Interpreter::Process<Cmd("meshdiag")>(Arg aArgs[])
+     * - RSS (average and last)
+     * - Error rates: frame tx (at MAC layer), IPv6 message tx (above MAC)
+     * - Connection time (seconds since link establishment {dd}d.{hh}:{mm}:{ss} format)
+     * - CSL info
```
Add colon to end

CSL info:

abtink commented on include/openthread/mesh_diag.h:
```diff
@@ -221,6 +229,181 @@ otError otMeshDiagGetNextIp6Address(otMeshDiagIp6AddrIterator *aIterator, otIp6A
+ * `mMessageErrorRate` values are valid. The frame error rate tracks frame tx errors (towards the child) at MAC
+ * layer,  while `mMessageErrorRate` tracks the IPv6 message error rate (above MAC layer and after MAC retries) when
+ * an IPv6 message is dropped. For example, if the message is large and requires 6LoWPAN fragmentation, message tx is
+ * considered as failed if one of its fragment frame tx fails (e.g., never acked).
```
Changed this and the rest in the new pushed commit.

Vyrastas commented on include/openthread/mesh_diag.h:
```diff
@@ -91,7 +91,7 @@ typedef struct otMeshDiagChildIterator otMeshDiagChildIterator;
 
 /**
- * Represents information about a router in Thread mesh.
+ * Represents information about a router in Thread mesh discovered using `otMeshDiagDiscoverTopology()`
```
I may have missed this one previously, missing period at end

abtink replied:
Fixed in new push. Thanks.

Vyrastas reviewed (approved):
LGTM, just one last nit
